UNIXFSv1.md

```diff
@@ -0,0 +1,487 @@
+
+This allows us to concatenate smaller files together.
+
+Linked files would be loaded recursively with the same process following a DFS (Depth-First-Search) order.
+
+Childs nodes must be of type file (so `Dag-PB` where type is `File` or `Raw`)
+
+For example this example pseudo-json block:
+```json
+{
+	"Links": [{"Hash":"Qmfoo"}, {"Hash":"Qmbar"}],
+	"Data": {
+		"Type": "File",
+		"blocksizes": [20, 30]
+	}
+}
+```
+
+This indicates that this file is the concatenation of the `Qmfoo` and `Qmbar` files.
+
+So when reading this file,
+the `blocksizes` array give us the size in bytes of the content of the child files, each index in `blocksizes` give the value at the same index in `Links`.
+
+This allows to do fast indexing into the file, for example if someone is trying to read bytes 25 to 35 we can compute an offset list by summing all previous indexes in `blocksizes`, then do a search to find which indexes contain the range we are intrested in.
+
+For example here the offset list would be `[0, 20]` and thus we know we only need to download `Qmbar` to get the range we are intrested in.
+
+If `blocksizes` or `Links` are not of the same length you MUST error.
+
+####### `decodeMessage(PBNode.Data).Data`
+
+This field is an array of bytes, it is file content and is appended before the links.
+
+This must be taken into a count when doing offsets calculations (the len of the `Data.Data` field define the value of the zeroth element of the offset list when computing offsets).
+
+####### Offset list
+
+The offset list isn't the only way to use blocksizes and reach a correct implementation, it is a simple cannonical one, python pseudo code to compute it looks like this:
+```python
+def offsetlist(node):
+	unixfs = decodeDataField(node.Data)
+	if len(node.Links) != len(unixfs.Blocksizes):
+		raise "unmatched sister lists" # error messages are implementation details
+
+	cursor = len(unixfs.Data) if unixfs.Data else 0
+	l = [] * len(node.Links)
+	for i, size in enumerate(unixfs.Blocksizes):
+		l[i] = cursor
+		cursor += size
+
+	return l
+```
+
+This will tell you which offset inside this node the children at the corresponding index starts to cover.
+
+####### `PBNode.Links[].Name` with Files
+
+They SHOULD be absent from the protobuf message, however for historic reasons we allow empty file names.
+
+If this field is present and non empty the file is invalid and you MUST error.
+
+####### `Blocksize` of a dag-pb file
+
+This is not a field present in the block directly, but rather a computable property of a file which would be used in parent files.
+It is the sum of the length of the `Data.Data` field plus the sum of all link's blocksizes.
+
+####### `PBNode.Data.Filesize`
+
+If present, this field must be equal to the `Blocksize` computation above, else the file is invalid.
+For now, the usage of this field is unknown given this could be computed from other fields, but some implementations emit it.
```

Comment:
Transmiting ~4bytes on each file roots in order to avoid a loop and some additions ?

That a surprising tradeoff to me but anyway I've removed this sentence.